### PR TITLE
Support the ability to use email address as login userid.

### PR DESCRIPTION
**Pull request by bcholmes**
Allow a user to user their email address, rather than their badge id (which they might not remember) to log in to Zambia

### CODE REVIEW

File: webpages/doLogin.php [hunk 1]
```diff
@@ -7,7 +7,10 @@
     $title = "Submit Password";
     $badgeid = getString('badgeid');
     $password = getString('passwd');
-    $query = "SELECT password, data_retention FROM Participants WHERE badgeid = ?;";
+    $query = "SELECT p.password, p.data_retention, p.badgeid FROM Participants p WHERE p.badgeid = ?;";
+    if (strpos($badgeid, '@') !== false && defined('EMAIL_LOGIN_SUPPORTED') && EMAIL_LOGIN_SUPPORTED === true) {
+        $query = "SELECT p.password, p.data_retention, p.badgeid FROM Participants p, CongoDump c WHERE p.badgeid = c.badgeid and c.email = ?;";
```
Comment (olszowka) [inline webpages/doLogin.php:12]: Please use the join syntax 
```
SELECT
        P.password, P.data_retention, P.badgeid
    FROM
             Participants P
        JOIN CongoDump C USING (badgeid)
    WHERE
        C.email = ?;
```
throughout Zambia.  As many as possibly of the foreign keys use identical names for the two columns in the two tables so this syntax may be used.  Where not, use the `ON` syntax.  I like to keep the join expressions out of the `WHERE` clause.  Also, use capital letters for table aliases.

**Review by olszowka — CHANGES_REQUESTED** [review]
There should be some prompt or advice to the user on the login screen that email address may be used instead of badgeid/prompt id.  Also, we need to document that `@` may not appear in a badgeid if email logins are permitted.

File: webpages/doLogin.php (hunk 1 above)
Comment (bcholmes) [reply webpages/doLogin.php:12]: Will do. I've also had a conversation with Syd about whether or not checking against '@' is 'safe' (as in 'safe to assume that a badgeid that has a @ in it must be an email address') and he made some additional recommendations. I ultimately came up with this query, instead (but haven't yet pushed it):

`SELECT p.password, p.data_retention, p.badgeid FROM Participants p WHERE p.badgeid = ? or p.badgeid in (select c.badgeid from CongoDump c where c.email = ?`

which I think handles the case where there's a record in Partipants, but not in CongoDump.

**Comment by olszowka** [comment]
I'm afraid the new login form has accessibility issues because screen readers look for form labels and don't necessarily read out placeholders and titles.  You can add labels with class "sr-only" so they won't be visible on the screen, but will be read by screen readers.

**Comment by olszowka** [comment]
I'm concerned that the old and new login forms look so different and that the old one still appears on some pages--I haven't dug in to figure out which.

**Comment by olszowka** [comment]
I'm not quite sure why you compare USER_ID_PROMPT to those two strings before appending " or Email address"?  It wouldn't really work for a con to set USER_ID_PROMPT to include something about email address already because there are other parts of the code where it is used and it wouldn't make sense.

**Comment by olszowka** [comment]
I'm not sure, but I think I want to have another option whether the password reset requires entering badgeid or not.

**Comment by bcholmes** [comment]
> I'm afraid the new login form has accessibility issues 

Excellent catch. I've fixed this in commit 0d7276f.

**Comment by bcholmes** [comment]
> I'm not quite sure why you compare USER_ID_PROMPT to those two strings before appending... 

In my use case, our members don't really know their badgeids, so I wanted to be able to support a case where the prompt simply says "Email address", but it's now clear to me that my approach to this has broader implications. I've pushed what I think is a better solution for this in fbee7ac .

**Pull request closed without merging**